pkg/scale/encode.go

```diff
@@ -0,0 +1,420 @@
+		// error case
+		err = es.WriteByte(1)
+		if err != nil {
+			break
+		}
+		// TODO: type checking of res.Err against resultCache to ensure Err is same as
+		// registered type
+		err = es.marshal(res.Err)
+	default:
+		err = es.WriteByte(0)
+		if err != nil {
+			break
+		}
+		// TODO: type checking of res.Ok against resultCache to ensure Err is same as
+		// registered type
+		err = es.marshal(res.Ok)
+	}
+	return
+}
+
+func (es *encodeState) encodeVaryingDataType(values VaryingDataType) (err error) {
+	err = es.encodeLength(len(values))
+	if err != nil {
+		return
+	}
+	for _, val := range values {
+		// TODO: type checking of val against vdtCache to ensure it is a registered type
+		// encode type.Index (idx) for varying data type
+		err = es.WriteByte(byte(val.Index()))
+		if err != nil {
+			return
+		}
+		err = es.marshal(val)
+	}
+	return
+}
+
+func (es *encodeState) encodeSlice(t interface{}) (err error) {
+	switch arr := t.(type) {
+	// int is the only case that handles encoding differently.
+	// all other cases can recursively call es.marshal()
+	case []int:
+		err = es.encodeLength(len(arr))
+		if err != nil {
+			return
+		}
+		for _, elem := range arr {
+			err = es.encodeUint(uint(elem))
+			if err != nil {
+				return
+			}
+		}
+	// the cases below are to avoid using the reflect library
+	// for commonly used slices in gossamer
+	case []*big.Int:
+		err = es.encodeLength(len(arr))
+		if err != nil {
+			return
+		}
+		for _, elem := range arr {
+			err = es.marshal(elem)
+			if err != nil {
+				return
+			}
+		}
+	case []bool:
+		err = es.encodeLength(len(arr))
+		if err != nil {
+			return
+		}
+		for _, elem := range arr {
+			err = es.marshal(elem)
+			if err != nil {
+				return
+			}
+		}
+	case [][]byte:
+		err = es.encodeLength(len(arr))
+		if err != nil {
+			return
+		}
+		for _, elem := range arr {
+			err = es.marshal(elem)
+			if err != nil {
+				return
+			}
+		}
+	case [][]int:
+		err = es.encodeLength(len(arr))
+		if err != nil {
+			return
+		}
+		for _, elem := range arr {
+			err = es.marshal(elem)
+			if err != nil {
+				return
+			}
+		}
+	case []string:
+		err = es.encodeLength(len(arr))
+		if err != nil {
+			return
+		}
+		for _, elem := range arr {
+			err = es.marshal(elem)
+			if err != nil {
+				return
+			}
+		}
+	default:
+		// use reflect package for any other cases
+		s := reflect.ValueOf(t)
+		err = es.encodeUint(uint(s.Len()))
+		if err != nil {
+			return
+		}
+		for i := 0; i < s.Len(); i++ {
+			err = es.marshal(s.Index(i).Interface())
+			if err != nil {
+				return
+			}
+		}
+	}
+	return
+}
+
+// encodeArray encodes an interface where the underlying type is an array
+// it writes the encoded length of the Array to the Encoder, then encodes and writes each value in the Array
+func (es *encodeState) encodeArray(in interface{}) (err error) {
+	v := reflect.ValueOf(in)
+	for i := 0; i < v.Len(); i++ {
+		elem := v.Index(i).Interface()
+		switch elem := elem.(type) {
+		case int:
```

Comment:
should this case have `int` and `uint`? I guess I'm unsure of the difference between this case and the default

Reply:
removed this case, no longer needed.